Neos.Flow/Classes/Cli/Dispatcher.php

```diff
@@ -0,0 +1,138 @@
+use Neos\Flow\Mvc\Controller\Exception\InvalidControllerException;
+use Neos\Flow\Mvc\Exception\ForwardException;
+use Neos\Flow\Mvc\Exception\InfiniteLoopException;
+use Neos\Flow\Mvc\Exception\StopActionException;
+use Neos\Flow\Mvc\Exception\UnsupportedRequestTypeException;
+use Neos\Flow\ObjectManagement\ObjectManagerInterface;
+
+/**
+ * @Flow\Scope("singleton")
+ */
+class Dispatcher
+{
+    /**
+     * @var \Neos\Flow\SignalSlot\Dispatcher
+     */
+    protected $signalDispatcher;
+
+    /**
+     * @var ObjectManagerInterface
+     */
+    protected $objectManager;
+
+    /**
+     * @param \Neos\Flow\SignalSlot\Dispatcher $signalDispatcher
+     */
+    public function injectSignalDispatcher(\Neos\Flow\SignalSlot\Dispatcher $signalDispatcher)
+    {
+        $this->signalDispatcher = $signalDispatcher;
+    }
+
+    /**
+     * @param ObjectManagerInterface $objectManager
+     */
+    public function injectObjectManager(ObjectManagerInterface $objectManager)
+    {
+        $this->objectManager = $objectManager;
+    }
+
+    /**
+     * Try processing the request until it is successfully marked "dispatched"
+     *
+     * @param Request $request
+     * @param Response $response
+     * @return Response
+     * @throws InvalidControllerException|InfiniteLoopException|NoSuchOptionException|UnsupportedRequestTypeException
+     */
+    public function dispatch(Request $request, Response $response)
+    {
+        $dispatchLoopCount = 0;
+        /** @var Request $request */
+        while (!$request->isDispatched()) {
+            if ($dispatchLoopCount++ > 99) {
+                throw new InfiniteLoopException(sprintf('Could not ultimately dispatch the request after %d iterations.', $dispatchLoopCount), 1217839467);
+            }
+            $controller = $this->resolveController($request);
+            try {
+                $this->emitBeforeControllerInvocation($request, $response, $controller);
+                $controller->processRequest($request, $response);
+                $this->emitAfterControllerInvocation($request, $response, $controller);
+            } catch (StopActionException $exception) {
+                $this->emitAfterControllerInvocation($request, $response, $controller);
+                if ($exception instanceof ForwardException) {
+                    $request = $exception->getNextRequest();
+                } elseif (!$request->isMainRequest()) {
+                    $request = $request->getParentRequest();
+                }
+            }
+        }
+
+        return $response;
+    }
+
+    /**
+     * This signal is emitted directly before the request is been dispatched to a controller.
+     *
+     * @param Request $request
+     * @param Response $response
+     * @param ControllerInterface $controller
+     * @return void
+     */
+    protected function emitBeforeControllerInvocation(Request $request, Response $response, ControllerInterface $controller)
```

Comment:
We also need to do a final iteration on these signals and decide how we want to proceed in the future. I think it woudl be nice if there was two distinct signals, one for any controller invocation and one specific to CLI and MVC.

Reply:
Ok, so the dispatchers irk me a bit. I kind of want an interface for dispatchers now, but the handle method arguments are different and not reconcilable in a way (the reason why we have two dispatchers now). I am thinking, the dispatcher is semi internal anyways, it's not api, so I am thinking, what if we declared this interface (not super helpful but at least something):
    
    /**
     * @return mixed some response to the handled request, depending on the implementation
     */
    public function handle();

And then instead of making the dispatcher implementations singleton, just have a constructor which gets the request, response (maybe even optional) and possible dependencies. The constructor would not be part of the interface then but the interface can point to the fact that you need to create implementations with some input (request).

That would allow us to have a signal on this interface and request implementations to emit a signal for the interface additionally to any other signals it emits.